### PR TITLE
Issue 4: Extracting email Domain

### PR DESCRIPTION
**Pull request by amittiwari1979**
@Stephen986 Not a PR, but since this is a first task, probably if you can review while i am working and any suggestions if you have. Still not done with this task as I still need to think this thru for default behaviour of adaptor.

### CODE REVIEW

File: seeds/email.csv [hunk 1]
```diff
@@ -0,0 +1,4 @@
+email
```
Comment (7mOatpfIPk6fXjlHH925mn3KecVmtlxoNDCVUpP) [inline seeds/email.csv]: Let's make the seed file name more specific, like email_addresses.  There could be another seed file with e-mail companies (Mailchimp, Klaviyo, Sailthru, etc.)

File: macros/extract_email_domain.sql [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{% macro extract_email_domain(email) -%}
+    {{ adapter.dispatch('extract_email_domain', 'census_utils') (email) }}
+{%- endmacro %}
+
+{% macro default__extract_email_domain(email) -%}
+    regexp_substr(lower(replace(rtrim({{ email }},'.'),' ','')), '@(.*)', 1, 1)
+{%- endmacro %}
+
+{%- macro snowflake__extract_email_domain(email) %}
+    regexp_substr(lower(replace(rtrim({{ email }},'.'),' ')), '@(.*)', 1, 1, 'e',1)
```
Comment (7mOatpfIPk6fXjlHH925mn3KecVmtlxoNDCVUpP) [inline macros/extract_email_domain.sql]: Oh yeah I noticed that the Snowflake version takes more parameters...

**Comment by 7mOatpfIPk6fXjlHH925mn3KecVmtlxoNDCVUpP** [comment]
Looks good, but needs handling for improperly formatted e-mails.  We should include some in the seed file like:

stephen ebrey@gmail.com
stephenebrey@g.mail.com
stephenebrey@gmailcom

**Comment by 8WlHSebsMM4Z71kjkvD2CAWSP7qyoVEA4ow6GLw** [comment]
> 

@Stephen986 I was thinking of how to go ahead with handling of  improperly formatted email. For example, in the below cases:

1. stephen ebrey@gmail.com : the domain is valid, but the email identifier is has a space. Since this is about extracting the domain, I am thinking that the overall email cleanup should be a separate macro.
2. stephenebrey@g.mail.com: Having 2 dots in the email domain is also valid. However, since gmail is a popular one, we can go ahead and fix it. Should we do a lookup against the free email provider and do the clean up of what ever domain we can? Or a more generic approach using the valid list of extensions like .com, .co.in, .co.au and keep that part and remove any invalid characters from domain which are not allowed as part of the domain.
3. stephenebrey@gmailcom: The approach here would be to do a lookup for the extension in the domain part not having an extension and then replace if found.
4. stephenebrey@yahoo : This is one case for example, can be yahoo.com or yahoo.co.au. How should be deal with this, does the macro return nothing in this case?

I will think more on the scenarios and work. Can we have a scheduled meeting may be twice a week for at-least 15-30 mins?

**Comment by 7mOatpfIPk6fXjlHH925mn3KecVmtlxoNDCVUpP** [comment]
> @Stephen986 I was thinking of how to go ahead with handling of improperly formatted email. For example, in the below cases:
> 
> 1. stephen [ebrey@gmail.com](mailto:ebrey@gmail.com) : the domain is valid, but the email identifier is has a space. Since this is about extracting the domain, I am thinking that the overall email cleanup should be a separate macro.

**True, that could be a separate macro, as long as this macro could still extract e-mail domain from an e-mail with spaces.**
> 2. [stephenebrey@g.mail.com](mailto:stephenebrey@g.mail.com): Having 2 dots in the email domain is also valid. However, since gmail is a popular one, we can go ahead and fix it. Should we do a lookup against the free email provider and do the clean up of what ever domain we can? Or a more generic approach using the valid list of extensions like .com, .co.in, .co.au and keep that part and remove any invalid characters from domain which are not allowed as part of the domain.

**Good point, I forgot that multiple periods is valid, and common for emails like @yahoo.co.jp.  It might be best to not do anything to validate then... most macros I've seen prioritize simplicity (so developers can easily understand what the macro is doing) over handling every edge case.  For example, this macro for getting URL parameters does not have any special validation for if someone accidentally put 'medium' instead of 'utm_medium' in a URL.  I think our macros might be hard to read if we include too much focus on edge cases.**
> 3. stephenebrey@gmailcom: The approach here would be to do a lookup for the extension in the domain part not having an extension and then replace if found.

**Yes, I like the idea of replacing 'com', 'org', 'net' etc with .com, .org, and .net if there is no . anywhere in the domain and they are at the end of the email.**
> 4. stephenebrey@yahoo : This is one case for example, can be yahoo.com or yahoo.co.au. How should be deal with this, does the macro return nothing in this case?

**We could add a ".com" to emails that don't have any extension at the end?  However, one thing is that even if we're fixing these in the DW, it won't be much use unless someone flags them and fixes them in other sources that e-mail the user.  So we should maybe also have a 'validate e-mail' macro that returns if the email is valid.  Then employees could look at the invalid emails and decide how and when to update them in other systems.  I'll discuss with Jeff at Census tomorrow.**

> I will think more on the scenarios and work. Can we have a scheduled meeting may be twice a week for at-least 15-30 mins?

**Yes, I set one up for Tuesday and Thursday morning.  Let me know if that does not work!**

File: seeds/email.csv (hunk 1 above)
Comment (8WlHSebsMM4Z71kjkvD2CAWSP7qyoVEA4ow6GLw) [reply seeds/email.csv]: This is done. As per suggestion, i have also renamed the seed file to prefix census_

**Comment by 8WlHSebsMM4Z71kjkvD2CAWSP7qyoVEA4ow6GLw** [comment]
> > @Stephen986 I was thinking of how to go ahead with handling of improperly formatted email. For example, in the below cases:
> > 
> > 1. stephen [ebrey@gmail.com](mailto:ebrey@gmail.com) : the domain is valid, but the email identifier is has a space. Since this is about extracting the domain, I am thinking that the overall email cleanup should be a separate macro.
> 
> **True, that could be a separate macro, as long as this macro could still extract e-mail domain from an e-mail with spaces.**
> 
> > 2. [stephenebrey@g.mail.com](mailto:stephenebrey@g.mail.com): Having 2 dots in the email domain is also valid. However, since gmail is a popular one, we can go ahead and fix it. Should we do a lookup against the free email provider and do the clean up of what ever domain we can? Or a more generic approach using the valid list of extensions like .com, .co.in, .co.au and keep that part and remove any invalid characters from domain which are not allowed as part of the domain.
> 
> **Good point, I forgot that multiple periods is valid, and common for emails like @yahoo.co.jp. It might be best to not do anything to validate then... most macros I've seen prioritize simplicity (so developers can easily understand what the macro is doing) over handling every edge case. For example, this macro for getting URL parameters does not have any special validation for if someone accidentally put 'medium' instead of 'utm_medium' in a URL. I think our macros might be hard to read if we include too much focus on edge cases.**
> 
> > 3. stephenebrey@gmailcom: The approach here would be to do a lookup for the extension in the domain part not having an extension and then replace if found.
> 
> **Yes, I like the idea of replacing 'com', 'org', 'net' etc with .com, .org, and .net if there is no . anywhere in the domain and they are at the end of the email.**
> 
> > 4. stephenebrey@yahoo : This is one case for example, can be yahoo.com or yahoo.co.au. How should be deal with this, does the macro return nothing in this case?
> 
> **We could add a ".com" to emails that don't have any extension at the end? However, one thing is that even if we're fixing these in the DW, it won't be much use unless someone flags them and fixes them in other sources that e-mail the user. So we should maybe also have a 'validate e-mail' macro that returns if the email is valid. Then employees could look at the invalid emails and decide how and when to update them in other systems. I'll discuss with Jeff at Census tomorrow.**
> 
> > I will think more on the scenarios and work. Can we have a scheduled meeting may be twice a week for at-least 15-30 mins?
> 
> **Yes, I set one up for Tuesday and Thursday morning. Let me know if that does not work!**
@Stephen986 

I have made changes for bigquery to handle the domain not having a period to handle popular extensions. I have used a key value pair dict to do the same. Please review and suggest your thoughts, so I will replicate the same for snowflake as well. I will discuss the reasons for not going to the dbt seed file instead of storing the values directly in dictionary.

Please let me know your thoughts on this.

**Comment by 7mOatpfIPk6fXjlHH925mn3KecVmtlxoNDCVUpP** [comment]
@amittiwari1979 can you move the tests from models/example to integration_tests/models and add a seed file for the expected results?

**Comment by 8WlHSebsMM4Z71kjkvD2CAWSP7qyoVEA4ow6GLw** [comment]
> @amittiwari1979 can you move the tests from models/example to integration_tests/models and add a seed file for the expected results?

@Stephen986 Moved the file and created integration test. Please review and let me know if it looks good to be merged or if you have any feedback.

File: integration_tests/models/test_extract_email_domain.sql [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{{ config(materialized="table") }}
+
```
Comment (7mOatpfIPk6fXjlHH925mn3KecVmtlxoNDCVUpP) [inline integration_tests/models/test_extract_email_domain.sql:2]: @amittiwari1979 dbt does not specify a materialization for their tests (letting them default to views), as you can see here: https://github.com/dbt-labs/dbt-utils/blob/1.0.0/integration_tests/models/sql/test_generate_surrogate_key.sql

I think we should follow their lead to be consistent?

**Review by Stephen986 — COMMENTED** [review]
Let's change the tests to views unless you think there's a good reason to do differently.

**Review by Stephen986 — APPROVED** [review]

**Pull request merged**